### PR TITLE
Change signature II: Handle arrow functions 

### PR DESCRIPTION
**Pull request by 11joselu**
Hey :) 

**This PR depends on https://github.com/nicoespeon/abracadabra/pull/746**

I was able to change function signature of an arrow function, but there are some doubts when to trigger "Change signature" action.

In this iteration we shoud put the Cursor on function asignation:

Untitled-2022-02-17-2107

But maybe we want to trigger "Change signature" over variable declaration. 

here-trigger

**Where do you think we should trigger it** @nicoespeon ?

### CODE REVIEW

**Comment by nicoespeon** [comment]
@11joselu I think it's fine to restrict the scope of the trigger to the function params itself for now 👍 
We'll see from usage if it feels like it would be intuitive to have it on the variable assignment too. But I think it's fair to put your cursor on the params if that's what you want to change.

File: src/refactorings/change-signature/change-signature.test.ts [hunk 1]
```diff
@@ -14,7 +14,7 @@ describe("Change Signature", () => {
     code: Code;
     expected: Code;
   }>(
-    "In same file",
+    "In same file with a function declarations",
```
Comment (nicoespeon) [inline src/refactorings/change-signature/change-signature.test.ts:17]:
```suggestion
    "In the same file with function declarations",
```

File: src/refactorings/change-signature/change-signature.test.ts [hunk 2]
```diff
@@ -185,6 +194,67 @@ describe("Change Signature", () => {
     }
   );
 
+  testEach<{
+    code: Code;
+    expected: Code;
+  }>(
+    "In same file with a arrow function declarations",
```
Comment (nicoespeon) [inline src/refactorings/change-signature/change-signature.test.ts:201]:
```suggestion
    "In the same file with arrow function declarations",
```

File: src/refactorings/change-signature/change-signature.test.ts [hunk 3]
```diff
@@ -25,6 +25,15 @@ describe("Change Signature", () => {
             return a + b;
           }`
       },
+      {
+        description: "when there are a arrow function",
```
Comment (nicoespeon) [inline src/refactorings/change-signature/change-signature.test.ts:29]: Shouldn't this test be moved to the new "testEach" that is about arrow functions?

File: src/refactorings/change-signature/change-signature.test.ts [hunk 4]
```diff
@@ -185,6 +194,67 @@ describe("Change Signature", () => {
...
   );
 
+  testEach<{
+    code: Code;
+    expected: Code;
+  }>(
+    "In same file with a arrow function declarations",
+    [
+      {
+        description: "when there are an arrow function without references",
```
Comment (nicoespeon) [inline src/refactorings/change-signature/change-signature.test.ts:204]:
```suggestion
        description: "when there is an arrow function without references",
```

File: src/refactorings/change-signature/change-signature.test.ts [hunk 5]
```diff
@@ -185,6 +194,67 @@ describe("Change Signature", () => {
...
+        code: `const add = [cursor](a, b) => {
+            return a + b;
+          }`,
+        expected: `const add = (b, a) => {
+            return a + b;
+          }`
+      },
+      {
+        description: "when there are an arrow function with references",
```
Comment (nicoespeon) [inline src/refactorings/change-signature/change-signature.test.ts:213]:
```suggestion
        description: "when there is an arrow function with references",
```

File: src/refactorings/change-signature/change-signature.test.ts [hunk 6]
```diff
@@ -185,6 +194,67 @@ describe("Change Signature", () => {
...
+          `,
+        expected: `const add = (b, a) => {
+            return a + b;
+          }
+          add(2, 1);
+          `
+      },
+      {
+        description:
+          "when there are a defined arrow function in multiple lines with references",
```
Comment (nicoespeon) [inline src/refactorings/change-signature/change-signature.test.ts:227]:
```suggestion
          "when there is a defined arrow function in multiple lines with references",
```

File: src/refactorings/change-signature/change-signature.ts [hunk 7]
```diff
@@ -135,13 +144,32 @@ function updateCode(
...
+      if (!t.isVariableDeclarator(path.parent)) return;
+
+      if (!path.parent.loc) return;
+
+      // For code Line always starts at 1. So we need to start at 0.
+      // Starting at 1 will get body function instead of function declaration
+      const arrowSelection = new Selection(
+        [path.parent.loc.start.line - 1, path.parent.loc.start.column],
+        [path.parent.loc.end.line - 1, path.parent.loc.end.column]
+      );
```
Comment (nicoespeon) [inline src/refactorings/change-signature/change-signature.ts:170]: 🍰 I think you should be able to use `Selection.fromAST(path.parent.loc)` for that. It handles the line start at 1 and all, so you can drop the comment in this case.

To be confirmed with tests, but that is a common issue that I solved in the past IMO.

**Comment by nicoespeon** [comment]
@11joselu could you update this PR with the latest main? It includes part 1 + a little change in the contract test that causes a conflict.

**Comment by 11joselu** [comment]
> @11joselu could you update this PR with the latest main? It includes part 1 + a little change in the contract test that causes a conflict.

Branch updated :)

I will update Class branch too

File: src/refactorings/change-signature/change-signature.ts (hunk 7 above)
Comment (11joselu) [reply src/refactorings/change-signature/change-signature.ts]: Oh! Nice factory.

Used at https://github.com/nicoespeon/abracadabra/commit/d3a78c1001b98783c49aef7234faef6160159ca4
✅ Test OK!

File: src/refactorings/change-signature/change-signature.test.ts (hunk 3 above)
Comment (11joselu) [reply src/refactorings/change-signature/change-signature.test.ts]: Is a duplicated test. 

Removed at https://github.com/nicoespeon/abracadabra/pull/747/commits/3579f40d0279cb13544161575aabacf263c42ca2

Nice catch!

**Pull request merged**